Spring Boot practice 1.0

aksoli666 opened the pull request:
added entity Category
added repository, service and controller
added liquibase

aksoli666 committed:
Spring Boot practice 1.0
added entity Category
added repository, service and controller
added liquibase

aksoli666 committed:
Spring Boot practice 1.0
improved solution and fixed search() method

liliia-ponomarenko commented on src/main/java/shop/controller/CategoryController.java:
```diff
@@ -0,0 +1,77 @@
+            summary = "Create a new category",
+            description = "Create a new category")
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @PostMapping
```
ResponseStatus.CREATED

liliia-ponomarenko commented on src/main/java/shop/controller/CategoryController.java:
```diff
@@ -0,0 +1,77 @@
+    )
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @GetMapping
+    public List<CategoryDto> getAll(Pageable pageable) {
```
better return Page instead of List  for realization pagination

liliia-ponomarenko commented on src/main/java/shop/controller/CategoryController.java:
```diff
@@ -0,0 +1,77 @@
+    )
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @DeleteMapping("/{id}")
+    public void deleteCategory(@PathVariable Long id) {
```
ResponseStatus.NO_CONTENT

liliia-ponomarenko commented on src/main/java/shop/dto/request/book/CreateBookRequestDto.java:
```diff
@@ -21,4 +23,5 @@ public class CreateBookRequestDto {
     private BigDecimal price;
     private String description;
     private String coverImage;
+    private Set<Long> categoryIds = new HashSet<>();
```
NotEmpty

liliia-ponomarenko commented on src/main/java/shop/dto/request/book/UpdateBookRequestDto.java:
```diff
@@ -21,4 +23,5 @@ public class UpdateBookRequestDto {
     private BigDecimal price;
     private String description;
     private String coverImage;
+    private Set<Long> categoryIds = new HashSet<>();
```
the same

liliia-ponomarenko commented on src/main/java/shop/dto/request/category/UpdateCategoryRequestDto.java:
```diff
@@ -0,0 +1,11 @@
+@Getter
+@Setter
+public class UpdateCategoryRequestDto {
+    private String name;
+    private String description;
```
Let's use the same validators ;)

liliia-ponomarenko commented on src/main/java/shop/entity/Book.java:
```diff
@@ -35,6 +42,15 @@ public class Book {
+            joinColumns = @JoinColumn(name = "book_id"),
+            inverseJoinColumns = @JoinColumn(name = "category_id")
+    )
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
```
You don't have such methods in your class, but you can leave it ;)

liliia-ponomarenko commented on src/main/java/shop/mapper/CategoryMapper.java:
```diff
@@ -0,0 +1,25 @@
+
+@Mapper(config = MapperConfig.class)
+public interface CategoryMapper {
+    Category toCategory(CreateCategoryRequestDto requestDto);
+
+    Category toCategory(CategoryDto categoryDto);
+
+    CategoryDto toCategoryDto(Category category);
+
+    CategoryDto toCategoryDto(CreateCategoryRequestDto requestDto);
+
+    List<CategoryDto> toCategoryDtoList(List<Category> categoryList);
+
+    void updateCategory(UpdateCategoryRequestDto requestDto, @MappingTarget Category category);
```
Do you use all these methods?

liliia-ponomarenko commented on src/main/java/shop/service/impl/BookServiceImpl.java:
```diff
@@ -56,9 +65,21 @@ public void deleteBookById(Long id) {
+                bookRepository.findAllByCategoryId(categoryId));
+    }
+
+    private void fetchCategoriesAndSetToBook(Set<Long> categoryIds, Book book) {
+        Set<Category> categories = new HashSet<>(categoryRepository.findAllById(categoryIds));
+        book.setCategories(categories);
+    }
```
Better use getReferenceById (let’s google how it works) 
```
private Set<Category> getCategoriesSet(Set<Long> categoriesIds) {
        return categoriesIds.stream()
                .map(categoryRepository::getReferenceById)
                .collect(Collectors.toSet());
    }
 ```

liliia-ponomarenko commented on src/main/java/shop/service/impl/BookServiceImpl.java:
```diff
@@ -56,9 +65,21 @@ public void deleteBookById(Long id) {
-                .stream()
+        Page<Book> bookPage = bookRepository.findAll(
+                bookSpecificationBuilder.build(searchParamsDto), pageable);
+        return bookPage.getContent().stream()
```
Let’s return the page.

liliia-ponomarenko commented on src/main/java/shop/service/impl/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,53 @@
+        return categoryMapper.toCategoryDto(category);
+    }
+
+    @Override
+    public List<CategoryDto> findAll(Pageable pageable) {
+        return categoryMapper.toCategoryDtoList(
+                categoryRepository.findAll(pageable).getContent());
+    }
```
the same

liliia-ponomarenko commented on src/main/java/shop/service/impl/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,53 @@
+
+    @Override
+    public void deleteById(Long id) {
+        categoryRepository.deleteById(id);
```
Should you check if the category exists by ID?

liliia-ponomarenko reviewed (changes requested):
Great! Let’s improve your solution ;)

aksoli666 committed:
Spring Boot practice 2.0
used ResponseStatus for create and delete methods;
return Page<> not List<>;
added validation for dto`s;
cleaned some methods in mapper for book and category;
used getReferenceById;

If i deleted entity, then i cant find it by id

Elena-Bruyako commented on src/main/java/shop/controller/CategoryController.java:
```diff
@@ -0,0 +1,81 @@
+import shop.service.CategoryService;
+
+@RestController
+@RequestMapping(value = "/api/categories", produces = MediaType.APPLICATION_JSON_VALUE)
```
You can add `server.servlet.context-path=/api` to application.properties
```suggestion
@RequestMapping(value = "/categories", produces = MediaType.APPLICATION_JSON_VALUE)
```

Elena-Bruyako commented on src/main/java/shop/controller/CategoryController.java:
```diff
@@ -0,0 +1,81 @@
+            description = "Create a new category")
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @ResponseStatus(HttpStatus.CREATED)
+    @PostMapping()
```
```suggestion
    @PostMapping
```

Elena-Bruyako reviewed (changes requested):
see comments

aksoli666 committed:
Spring Boot practice 1.0
deleted context path
deleted brackets
